### ISSUE TITLE
Out-of-Bounds Index Error in Tensor Access using tf.gather

### ISSUE THREAD

**Issue by Camwalker9900** [calico/borzoi-paper#2]
Hello, 

I am trying to regenerate the results from figure 2a. 

I ran this command: 

python /mount/samee/cam/projects/borzoi/borzoi/borzoi_satg_gene_gpu.py -o gtex_muscle_log2fc_undo_clip -f 0,1,2,3 --rc 1 --shifts 0 --span 0 --smoothgrad 0 --clip_soft 384.0 --fa /mount/samee/cam/projects/borzoi/data/hg38.fa -t /mount/samee/cam/projects/borzoi/data/Fig_2a/targets/targets_blood.txt /mount/samee/cam/projects/borzoi/data/test01/params_pred.json /mount/samee/cam/projects/borzoi/borzoi /mount/samee/cam/projects/borzoi/data/test01/gtex_diff_expr_log2fc_5k.gtf 

I got this error: 

There are 4999 genes in the gene list.
clip_soft = 384.0
n genes = 4999
-- Fold = 0 --
 - (gradients) - 
Processing shift 0
Fwd/rev = fwd
Processing 0, ENSG00000000003.15
Processing 500, ENSG00000089169.15
Processing 1000, ENSG00000109906.15
2024-08-01 10:33:03.449247: I tensorflow/core/common_runtime/executor.cc:1197] [/job:localhost/replica:0/task:0/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: indices[0,89] = 7611 is not in [0, 7611)
	 [[{{node GatherV2}}]]
Traceback (most recent call last):
  File "/mount/samee/cam/projects/borzoi/borzoi/borzoi_satg_gene_gpu.py", line 964, in <module>
    main()
  File "/mount/samee/cam/projects/borzoi/borzoi/borzoi_satg_gene_gpu.py", line 652, in main
    grads = seqnn_model.gradients(
  File "/mount/samee/cam/projects/borzoi/baskerville/src/baskerville/seqnn.py", line 621, in gradients
    self.gradients_func(
  File "/mount/samee/cam/miniconda/envs/borzoi_env/lib/python3.9/site-packages/tensorflow/python/util/traceback_utils.py", line 153, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/mount/samee/cam/miniconda/envs/borzoi_env/lib/python3.9/site-packages/tensorflow/python/eager/execute.py", line 52, in quick_execute
    tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,
tensorflow.python.framework.errors_impl.InvalidArgumentError: Graph execution error:

Detected at node 'GatherV2' defined at (most recent call last):
    File "/mount/samee/cam/projects/borzoi/borzoi/borzoi_satg_gene_gpu.py", line 964, in <module>
      main()
    File "/mount/samee/cam/projects/borzoi/borzoi/borzoi_satg_gene_gpu.py", line 652, in main
      grads = seqnn_model.gradients(
    File "/mount/samee/cam/projects/borzoi/baskerville/src/baskerville/seqnn.py", line 621, in gradients
      self.gradients_func(
    File "/mount/samee/cam/projects/borzoi/baskerville/src/baskerville/seqnn.py", line 706, in gradients_func
      preds = tf.gather(
Node: 'GatherV2'
indices[0,89] = 7611 is not in [0, 7611)
	 [[{{node GatherV2}}]] [Op:__inference_gradients_func_145360]
(borzoi_env) [u253262@compute-9-0 borzoi]$ print("Tensor size:", tensor_to_gather_from.shape)
bash: syntax error near unexpected token `"Tensor size:",'
(borzoi_env) [u253262@compute-9-0 borzoi]$ print("Indices accessing:", indices_being_accessed)
bash: syntax error near unexpected token `"Indices accessing:",'

Thanks for the Help!

**Comment by johli** [comment]
Hi,

I am looking into this issue. Could you attach the "targets_blood.txt" and "params_pred.json" that you're using?

Best,

Johannes

**Comment by Camwalker9900** [comment]
Hi Johannes, 

I resolved a previous error by ensuring the values in the first column of my "targets_blood.txt" matched the corresponding values in each respective data file. However, after making this correction, the command continued to run for over 24 hours without completion. In an attempt to address this, I updated my GTF file to include only CFHR2, but this process is still ongoing. Given that I’m using an HPCC, I expected sufficient resources to handle this.

Could extended run times be typical under these circumstances? If so, what steps can I take to verify that the process is progressing and not stalled?

Additionally, I would appreciate guidance on how to create a custom targets.txt file compatible with the model using my own RNA-seq data.

Finally, for running borzoi_satg_gene_gpu.py on mouse data, would replacing the human GTF file with a mouse-specific GTF file suffice?

Thank you for your assistance, 
Cameron

[updated_targets_blood.txt](https://github.com/user-attachments/files/16530721/updated_targets_blood.txt)
The params_pred.json file I am using is the one from the examples directory in the Borzoi repository.
